pkg/scheduler/plugins/binpack/binpack.go

```diff
@@ -0,0 +1,230 @@
+		BinPackingMemory:    1,
+		BinPackingResources: make(map[v1.ResourceName]int),
+	}
+
+	// Checks whether binpack.weight is provided or not, if given, modifies the value in weight struct.
+	args.GetInt(&weight.BinPackingWeight, BinpackWeight)
+	// Checks whether binpack.cpu is provided or not, if given, modifies the value in weight struct.
+	args.GetInt(&weight.BinPackingCPU, BinpackCPU)
+	if weight.BinPackingCPU < 0 {
+		weight.BinPackingCPU = 1
+	}
+	// Checks whether binpack.memory is provided or not, if given, modifies the value in weight struct.
+	args.GetInt(&weight.BinPackingMemory, BinpackMemory)
+	if weight.BinPackingMemory < 0 {
+		weight.BinPackingMemory = 1
+	}
+
+	resourcesStr := args[BinpackResources]
+	resources := strings.Split(resourcesStr, ",")
+	for _, resource := range resources {
+		resource = strings.TrimSpace(resource)
+		if resource == "" {
+			continue
+		}
+
+		// binpack.resources.[ResourceName]
+		resourceKey := BinpackResourcesPrefix + resource
+		resourceWeight := 1
+		args.GetInt(&resourceWeight, resourceKey)
+		if resourceWeight < 0 {
+			resourceWeight = 1
+		}
+		weight.BinPackingResources[v1.ResourceName(resource)] = resourceWeight
+	}
+
+	return weight
+}
+
+func (bp *binpackPlugin) Name() string {
+	return PluginName
+}
+
+func (bp *binpackPlugin) OnSessionOpen(ssn *framework.Session) {
+	glog.V(3).Infof("Enter binpack plugin ...")
+	if glog.V(3) {
+		defer func() {
+			glog.V(3).Infof("Leaving binpack plugin. %s ...", bp.weight.String())
+		}()
+	}
+
+	nodeOrderFn := func(task *api.TaskInfo, node *api.NodeInfo) (float64, error) {
+		binPackingScore := BinPackingScore(task, node, bp.weight)
+
+		glog.V(4).Infof("Binpack score for Task %s/%s on node %s is: %v", task.Namespace, task.Name, node.Name, binPackingScore)
+		return binPackingScore, nil
+	}
+	if bp.weight.BinPackingWeight != 0 {
+		ssn.AddNodeOrderFn(bp.Name(), nodeOrderFn)
+	} else {
+		glog.Infof("binpack weight is zero, skip node order function")
+	}
+}
+
+func (bp *binpackPlugin) OnSessionClose(ssn *framework.Session) {
+}
+
+// BinPackingScore use the best fit polices during scheduling.
+// Goals:
+// - Schedule Jobs using BestFit Policy using Resource Bin Packing Priority Function
+// - Reduce Fragmentation of scarce resources on the Cluster
+func BinPackingScore(task *api.TaskInfo, node *api.NodeInfo, weight priorityWeight) float64 {
+	score := 0.0
+	weightSum := 0
+	requested := task.Resreq
+	allocatable := node.Allocatable
+	used := node.Used
+
+	score += ResourceBinPackingScore(requested.MilliCPU, allocatable.MilliCPU, used.MilliCPU, weight.BinPackingCPU)
+	weightSum += weight.BinPackingCPU
+	score += ResourceBinPackingScore(requested.Memory, allocatable.Memory, used.Memory, weight.BinPackingMemory)
+	weightSum += weight.BinPackingMemory
+
+	// All resource with weight should be calculated, because the weightSum need it,
+	// even the node have no this resource.
+	for name, weight := range weight.BinPackingResources {
+		weightSum += weight
```

Comment:
If score calculated at L202 is zero, whether we need to add `weight` to `weightSumt`?

Reply:
Zero is also a valid value for score, so I think add it to sum is also valid.

Reply:
Take GPU for example: 
Node A has no GPU, so its GPU score will be 0, Node B has GPU capacity 6 and use 3, then it has a score for GPU resource.  Suppose that Node A and B's other resource's scores are same, if we add `weight` in this case, Node B's total score will be higher than Node A, then it will choose Node B, is it expected? Node A might be a better choice?

Reply:
if the resource is not exist on the node, we should also skip weight for it.

Reply:
In order to choose NodeA, maybe we should make the GPU score in NodeB to be nagetive in this case?

For example, request need resource Ra/Rb, but the NodeA has resource Ra/Rb, NodeB has resource Ra/Rb/Rc.
Policy Current is the score now. Policy Skip will ignore the Sc/Wc in NodeA but could not guarantee that NodeA have a higher score than NodeB. Policy Negative would treat the Sc in NodeB as negative score.
```
policy      NodeA              NodeB
current     Sa+Sb+0/Wa+Wb+Wc   Sa+Sb+Sc/Wa+Wb+Wc
skip        Sa+Sb/Wa+Wb        Sa+Sb+Sc/Wa+Wb+Wc
negative    Sa+Sb+0/Wa+Wb+Wc   Sa+Sb-Sc/Wa+Wb+Wc
```

Reply:
I don't have any better idea for this now. Just some thought:

* If request does not need resource C, then we do not calculate the score of this resource, it means that we compare `Sa+Sb/Wa+Wb`(for NodeA)  with `Sa+Sb/Wa+Wb`(for Node B).

or 

* If NodeA does not has the resource, we regards its RequestedToCapacityRatio to 1 instead of 0, it means that it is full, 0 seems it is empty.

Reply:
Option 1 is better to me :)

Reply:
I refactor the code in `BinPackingScore`. Now, only resources that pod requested can affect the binpack score.